bug: arrays.map does not validate callback return type, produces garbage on type mismatch

## Description

`arrays.map()` does not validate that the mapping callback's return type matches the input array's element type. When the callback returns a different type, the typechecker accepts the call and codegen produces code that reinterprets the return value's bits as the input element type, resulting in garbage data at runtime.

`arrays.filter()`, `arrays.any()`, and `arrays.all()` correctly validate their callbacks with E9004. `arrays.map()` is missing the same check.

## Reproduction

```grayscale
import @arrays

do to_str(n int) -> string {
    return "n${n}"
}

do main() {
    mut nums [int] = {1, 2, 3}
    mut strs = arrays.map(nums, ()to_str)
    println(strs)
    println(type_of(strs))
}
```

## Expected

A compile-time error such as:

```
error[E9004]: arrays.map() callback signature mismatch; map callback must return the same type as the array element type (int)
```

## Actual

Compiles without error. Prints garbage at runtime:

```
{5234524296, 5234524304, 5234524312}
[int]
```

The string pointer values are reinterpreted as `int64`. `type_of()` reports `[int]` — the input type, not the callback's return type.

## Related

Related to #2241 — same missing callback validation in arrays.reduce().

--- a/grayc/src/typechecker/typechecker.c
+++ b/grayc/src/typechecker/typechecker.c
@@ -2622,6 +2622,15 @@ static GrayType *resolve_stdlib_call(TypeChecker *checker, AstNode *node, const 
                                     diagnostic_error_code_formatted(checker->diag, "E9004",
                                         NODE_FILE(checker, cb_arg), cb_arg->token.line, cb_arg->token.column, 0,
                                         mfn, msg);
+                                } else if (elem_tn && cb_fs->return_count >= 1 &&
+                                           cb_fs->return_types[0] &&
+                                           strcmp(type_name(cb_fs->return_types[0]), elem_tn) != 0) {
+                                    char *msg = typechecker_format(checker,
+                                        "map callback must return the same type as the array element type (%s), got '%s'",
+                                        elem_tn, type_name(cb_fs->return_types[0]));
+                                    diagnostic_error_code_formatted(checker->diag, "E9004",
+                                        NODE_FILE(checker, cb_arg), cb_arg->token.line, cb_arg->token.column, 0,
+                                        mfn, msg);
                                 }
                             } else if (strcmp(mfn, "filter") == 0 ||
                                        strcmp(mfn, "any") == 0 ||

--- a/integration-tests/fail/errors/E9004_map_callback_return_type.gray
+++ b/integration-tests/fail/errors/E9004_map_callback_return_type.gray
@@ -1,0 +1,15 @@
+/*
+ * Error Test: E9004 - arrays.map() callback return type mismatch
+ * Expected: E9004 error (map callback must return the same type as the array element type)
+ */
+
+import @arrays
+
+do to_str(n int) -> string {
+    return "n${n}"
+}
+
+do main() {
+    mut nums [int] = {1, 2, 3}
+    mut strs = arrays.map(nums, ()to_str)
+}

